proto/backend.proto

```diff
@@ -0,0 +1,217 @@
+
+message PluginRequest {
+  // Plugin Configuration
+  PluginConfig config = 1;
+
+  // Environment info
+  map<string,string> headers = 2;
+
+  // HTTP Style parameters
+  string method = 3; // GET/POST/DELETE/etc
+  string path = 4;   // (with params?)
+  bytes body = 5;    // assume string unless header says otherwise?
+}
+
+message PluginResponse {
+  int32 code = 1;  // Match HTTP status codes
+  bytes body = 2;
+  map<string,string> headers = 3;
+}
+
+
+
+//-----------------------------------------------
+// All Backend Plugins Implement
+//-----------------------------------------------
+
+service BackendPlugin {
+  // Get the plugin status
+  rpc Check(PluginStatusRequest) returns (PluginStatusResponse);
+
+  // HTTP Style request
+  rpc Request(PluginRequest) returns (PluginResponse);
+
+  // Well typed query interface
+  rpc Query(QueryRequest) returns (QueryResponse);
+
+  // TODO -- currently queries die when the response >4mb...
+  // We need a better solution...
+  // a) encode paging info in the DataFrame metadata?
+  // b) chunk the results and grafana backend would
+}
+
+
+//-----------------------------------------------
+// Allow plugins to access other plugins
+// Only allow this for some plugins?  GEL only
+//
+// ???? How does the backend know the PluginConfig????
+//-----------------------------------------------
+
+service GrafanaPlatform {
+  rpc PlatformPluginRequest(PluginRequest) returns (PluginResponse);
+
+  rpc PlatformPluginQuery(QueryRequest) returns (QueryResponse);
+}
+
+
+//-----------------------------------------------------------------
+// Streaming support aimed to support similar features to STOMP
+// consider Apache Flight for more complex/standard
+//-----------------------------------------------------------------
+
+message StreamingRequest {
+  int64 pluginId = 1; // match the id from connect request (app or datasource)
+
+  // Environment info
+  map<string,string> headers = 1;
+
+  // The topic we are talking to
+  string topic = 2;
+
+  // The message
+  string body = 3;
+
+  // Flag if the user is already listening
+  boolean subscribed = 4;
+}
+
+message StreamingMessage {
+  int64 pluginId = 1;
+  string topic = 2; // will be broadcast to eveyone listening to this app|ds + topic
+  string message = 3; // JSON Blob
+}
+
+service StreamingPlugin {
+  // Called by grafana before the first request
+  // Grafana server will set up a websocket that will broadcast to all listeners
+  rpc Connect(PluginConfig) returns (stream StreamingMessage);
+
+  // If the message is not empty, send that to the user
+  // throwing an error will make sure the user is not subscribed
+  rpc Process(StreamingRequest) returns (StreamingMessage)
+
+  // Stop streaming to that instance, returns the number of open connections
+  rpc Disconnect(int64) returns (int64);
+}
+
+
+//-----------------------------------------------------------------
+// Renderer (do we want this definition here?)
```

Comment:
I think renderer might be a bit different here. Consider the case of "datasource applications" for something like Graphite or OpenTSDB. They take a query, and return an image.  Or perhaps you query some satellite map api and get an image back.

There is also the Grafana Rendering server. This takes data (for example, time series data) and returns a Grafana Rendering of the image.

I think the Grafana Rendering service is a different think and likely doesn't belong in this protocol. 

I can see a use something like `rpc Query(QueryRequest) returns (ImageResponse)` for things like the satellite image example that could be here, but that would be an rpc call in the BackendPlugin service, and be a different response.